machine/src/namespace.rs

```diff
@@ -51,10 +52,33 @@ impl Namespace {
         }
         Ok(())
     }
+
+    /// Enter this namespace and change back to previous one when guard is dropped
+    pub fn enter_guarded(&self) -> Result<NamespaceGuard, io::Error> {
+        let prior_ns = Self::current()?;
+        self.enter()?;
+        Ok(NamespaceGuard {
+            prior_ns,
+            _ph: PhantomData,
+        })
```

Comment:
I don't think this will work as intended. You can enter a namespace given a pid/tid. But it seems each thread can own a namespace which may or may not be shared with other threads. So Namespace::current() is always the same even when entering a different threads namespace

Reply:
Also since we require futures to be send, this will be unusable in practice

Reply:
more precisely: switching namespaces is fundamentally incompatible with using futures, since futures live in a lexical scope that differs a lot from the execution context — after an `.await` it might run on a different thread or the current thread may have been modified

Reply:
In theory I agree in practice it only happens if actual work stealing is performed. That is why the recommended way to use netsim is via swarm-cli